src/main/kotlin/org/jmailen/gradle/kotlinter/KotlinterPlugin.kt

```diff
@@ -1,104 +1,84 @@
 package org.jmailen.gradle.kotlinter
 
 import com.android.build.gradle.BaseExtension
+import com.android.build.gradle.api.AndroidSourceSet
 import org.gradle.api.Plugin
 import org.gradle.api.Project
-import org.gradle.api.file.FileCollection
 import org.gradle.api.file.FileTree
 import org.gradle.api.file.SourceDirectorySet
 import org.gradle.api.internal.HasConvention
 import org.gradle.api.plugins.JavaPluginConvention
 import org.gradle.api.tasks.SourceSet
+import org.gradle.api.tasks.SourceSetContainer
 import org.jetbrains.kotlin.gradle.plugin.KotlinSourceSet
 import org.jmailen.gradle.kotlinter.support.reporterFileExtension
 import org.jmailen.gradle.kotlinter.tasks.FormatTask
 import org.jmailen.gradle.kotlinter.tasks.LintTask
+import java.io.File
 
 class KotlinterPlugin : Plugin<Project> {
 
     val extendablePlugins = mapOf(
-            "org.jetbrains.kotlin.jvm" to this::kotlinSourceSets,
-            "kotlin-android" to this::androidSourceSets)
+            "org.jetbrains.kotlin.jvm" to KotlinJvmSourceSetResolver,
+            "kotlin-android" to AndroidSourceSetResolver)
 
     override fun apply(project: Project) {
-        val kotlinterExtention = project.extensions.create("kotlinter", KotlinterExtension::class.java)
+        val kotlinterExtension = project.extensions.create("kotlinter", KotlinterExtension::class.java)
 
         // for known kotlin plugins, create tasks by convention.
-        val kotlinApplier = KotlinterApplier(project)
+        val taskCreator = TaskCreator(project)
         extendablePlugins.forEach { pluginId, sourceResolver ->
             project.plugins.withId(pluginId) {
-                val sourceSets = sourceResolver(project)
-                kotlinApplier.createTasks(sourceSets)
+
+                sourceResolver.applyToAll(project) { id, files ->
```

Comment:
It looks like source sets are resolved at application time, which means if a source set is added after this plugin is applied it won't pick up on it. I would consider using SourceSet's `whenObjectAdded { ... }` function to allow you to pick up source sets that are created long after this plugin is applied.

Reply:
I _think_ this would also be a pretty easy change that would not break any backwards compatibility (unlike my other comment).

Reply:
This refactor is subtle, but we no longer resolve source sets at apply time. We do it on the source set `all` as you wisely suggested we do.

Take a close read of `applyToAll` in the two implemenations `KotlinJvmSourceSetResolver` and `AndroidSourceSetResolver`. The second argument of `applyToAll` is a block to be called within `SourceSetContainer.all { ... }`. So the `TaskCreator` will do its thing once the source sets are resolved properly for the project.